lib/model/model.go

```diff
@@ -906,20 +912,71 @@ func (m *Model) ClusterConfig(deviceID protocol.DeviceID, cm protocol.ClusterCon
 				// We don't yet share this folder with this device. Add the device
 				// to sharing list of the folder.
 
-				l.Infof("Adding device %v to share %q (vouched for by introducer %v)", id, folder.ID, deviceID)
+				l.Infof("Sharing folder %q with %v (vouched for by introducer %v)", folder.ID, id, deviceID)
 
 				m.deviceFolders[id] = append(m.deviceFolders[id], folder.ID)
 				m.folderDevices[folder.ID] = append(m.folderDevices[folder.ID], id)
 
 				folderCfg := m.cfg.Folders()[folder.ID]
 				folderCfg.Devices = append(folderCfg.Devices, config.FolderDeviceConfiguration{
-					DeviceID: id,
+					DeviceID:     id,
+					IntroducedBy: deviceID,
 				})
 				m.cfg.SetFolder(folderCfg)
 
 				changed = true
 			}
 		}
+
+		if !devCfg.SkipIntroductionRemovals {
```

Comment:
This whole following section is large and complex and nested. I'd like to see it broken out into a couple of methods with good names and maybe a comment or two on what is going on?